pkg/operator/csidriveroperator/olmremovalcontroller.go

```diff
@@ -0,0 +1,463 @@
+	}
+
+	// 5. Delete the driver DaemonSet.
+	removed, err = c.ensureDriverDaemonSetRemoved(ctx, c.olmOptions.CSIDriverNamespace, c.olmOptions.CSIDriverDaemonSetName)
+	if err != nil {
+		return err
+	}
+	if !removed {
+		klog.V(4).Infof("CSIDriverMigrationController.Sync waiting for the driver DaemonSet to disappear")
+		return c.markProgressing("Waiting for OLM driver DaemonSet to be deleted")
+	}
+
+	// 6. Remove CR
+	removed, err = c.ensureCRRemoved(ctx, c.olmOptions.CRResource)
+	if err != nil {
+		return err
+	}
+	if !removed {
+		klog.V(4).Infof("OLMOperatorRemovalController.Sync waiting for the CR to disappear")
+		return c.markProgressing("Waiting for the old operator CR to be deleted")
+	}
+
+	klog.V(4).Infof("OLMOperatorRemovalController.Sync done!")
+	return c.markFinished("CSI driver has been removed from OLM")
+}
+
+func (c *OLMOperatorRemovalController) findSubscription(ctx context.Context) (string, string, string, bool, error) {
+	subscriptions, err := c.dynamicClient.Resource(subscriptionResourceGVR).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return "", "", "", false, err
+	}
+	for _, obj := range subscriptions.Items {
+		pkg, found, err := unstructured.NestedString(obj.Object, "spec", "name")
+		if !found {
+			continue
+		}
+		if err != nil {
+			return "", "", "", false, err
+		}
+		source, found, err := unstructured.NestedString(obj.Object, "spec", "source")
+		if !found {
+			continue
+		}
+		if err != nil {
+			return "", "", "", false, err
+		}
+		sourceNamespace, found, err := unstructured.NestedString(obj.Object, "spec", "sourceNamespace")
+		if !found {
+			continue
+		}
+		if err != nil {
+			return "", "", "", false, err
+		}
+
+		if pkg == c.olmOptions.OLMPackageName && source == olmSource && sourceNamespace == olmSourceNamespace {
+			csvName, _, err := unstructured.NestedString(obj.Object, "status", "currentCSV")
+			if err != nil {
+				return "", "", "", false, err
+			}
+			klog.V(2).Infof("Found subscription %s/%s with CSV %s", obj.GetNamespace(), obj.GetName(), csvName)
+			return obj.GetNamespace(), obj.GetName(), csvName, true, nil
+		}
+	}
+	return "", "", "", false, nil
+}
+
+func (c *OLMOperatorRemovalController) deleteSubscription(ctx context.Context, namespace, name string) (bool, error) {
+	err := c.dynamicClient.Resource(subscriptionResourceGVR).Namespace(namespace).Delete(ctx, name, metav1.DeleteOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+	klog.V(4).Infof("Deleted subscription %s/%s", namespace, name)
+	return false, nil
+}
+
+func (c *OLMOperatorRemovalController) deleteCSV(ctx context.Context, namespace, name string) (bool, error) {
+	err := c.dynamicClient.Resource(csvResourceGVR).Namespace(namespace).Delete(ctx, name, metav1.DeleteOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+	klog.V(4).Infof("Deleted CSV %s/%s", namespace, name)
+	return false, nil
```

Comment:
shouldn't you return true here?

Reply:
No, `Delete()` success does not mean the object is deleted. I added a comment there.